### ISSUE TITLE
TypeError thrown when decoding invalid value

### ISSUE THREAD

**Issue by Pudge601** [Spomky-Labs/base64url#11]
After upgrading to version 2, the use of strict typing now means that the `decode` method will throw a `TypeError` when given nonsense input.

Example:
```
\Base64Url\Base64Url::decode('cxr0fdsezrewklerewxoz423ocfsa3bw432yjydsa9lhdsalw');
```
This would previously have returned `false` as that is what `base64_decode` is returning, but with the strict type hint for returning a `string`, this now errors.

I'm not sure if you want to class this as an issue, as it could just be considered a change of behaviour (which I guess is okay since it's a change in a major version), but if this is intended, then perhaps there should be an `@throws TypeError` added to the docblock to state this explicitly.

Also, there doesn't seem to be any unit tests which test nonsense strings such as the one above, so that might be worth adding.

**Comment by Spomky** [comment]
Hi,

The previous behaviour was not in line with the docblock.
You was supposed to receive a string but for invalid inputs you received undocumented boolean (false).
Throwing an exception is the normal behaviour for invalid outputs e.i. invalid inputs. However, you are right to say that it should be mentioned in the docblock.

I will add a `@throws` block accordingly.

I prefer no to add a `if` statement as it will decrease performance just to throw an `\InvalidArgumentException` instead of a `\TypeError` one. The block I will add will be a generic one e.g. `@throws \Exception Thrown in case of invalid input`.

**Comment by Pudge601** [comment]
Hmm.. though I did suggest that, I would actually say that just allowing the code to throw a `TypeError` and document it is probably the least correct option to solve this.

Errors are not supposed to be thrown in the normal running of code - they're meant to indicate that the code is broken, which in this case, it is.

As for not wanting to add the `if` statement for performance reasons, I think that would perhaps the worst example of a micro-optimisation I've heard of (for the 99% of expected usage where the value will not be `false`, it would be so small an impact to be considered neglible). As it is, the code is currently broken - not fixing it because of the performance impact of a simple equality check seems very wrong to me.

**Comment by Spomky** [comment]
Makes sense.
Merged and tagged `v2.0.1`.
Thanks
